File: examples-java/lecture/src/main/java/hwr/oop/examples/classes/comparability/Train.java
package hwr.oop.examples.classes.comparability;

class Train implements Comparable<Train> {

  private final int numberOfRailCars;
  private int speed;

  public Train(int numberOfRailCars) {
    this.numberOfRailCars = numberOfRailCars;
    this.speed = 0;
  }

  @Override
  public int compareTo(Train o) {
    return Integer.compare(this.speed, o.speed);
  }

  void accelerateTo(int speed) {
    this.speed = speed;
  }

  int getRailCars() {
    return numberOfRailCars;
  }
}

File: examples-kotlin/lecture-04/src/main/kotlin/hwr/oop/examples/Train.kt
package hwr.oop.examples

class Train(
  private val name: String = "ICE1337",
  private val start: String = "Berlin",
  private val destination: String = "Hamburg",
  val numberOfWagons: Int = 1,
) : Comparable<Train> {

  private var speed: Int = 0

  fun accelerateTo(speed: Int) {
    this.speed = speed
  }

  override fun equals(other: Any?): Boolean {
    // ignores numberOfWagons and speed
    if (this === other) return true
    if (other == null) return false
    if (this::class != other::class) return false

    other as Train

    if (name != other.name) return false
    if (start != other.start) return false
    if (destination != other.destination) return false

    return true
  }

  override fun hashCode(): Int {
    var result = name.hashCode()
    result = 31 * result + start.hashCode()
    result = 31 * result + destination.hashCode()
    return result
  }

  override fun compareTo(other: Train): Int {
    return speed.compareTo(other.speed)
  }

}